<?xml version="1.0" encoding="UTF-8"?>

<mule
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xmlns:munit="http://www.mulesoft.org/schema/mule/munit"
  xmlns:munit-tools="http://www.mulesoft.org/schema/mule/munit-tools"
  xmlns="http://www.mulesoft.org/schema/mule/core"
  xmlns:doc="http://www.mulesoft.org/schema/mule/documentation"
  xsi:schemaLocation="
		http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
		http://www.mulesoft.org/schema/mule/munit http://www.mulesoft.org/schema/mule/munit/current/mule-munit.xsd
		http://www.mulesoft.org/schema/mule/munit-tools  http://www.mulesoft.org/schema/mule/munit-tools/current/mule-munit-tools.xsd">
  <munit:config name="impl-option-d-test-suite.xml" />


  <munit:test
    name="test-test-suite-testFlowTest"
    timeOut="9000000">
    <munit:enable-flow-sources>
      <munit:enable-flow-source
        value="munit-set-error-code-event-flow" />
    </munit:enable-flow-sources>
    <munit:behavior>
      <munit-tools:mock-when
        doc:name="Mock when Request -&gt; then call"
        processor="http:request">
        <munit-tools:with-attributes>
          <munit-tools:with-attribute
            whereValue="GET"
            attributeName="method" />
          <munit-tools:with-attribute
            whereValue="http://example.com/some-resource"
            attributeName="url" />
        </munit-tools:with-attributes>
        <munit-tools:then-call
          flow="munit-set-error-code-event-flow" />
      </munit-tools:mock-when>
    </munit:behavior>
    <munit:execution>
      <flow-ref
        doc:name="Flow-ref to testFlow"
        name="test2Flow" />
    </munit:execution>
    <munit:validation>
      <munit-tools:verify-call
        doc:name="Some logic for CODE_A"
        doc:id="c6f7ec31-39dc-4e5b-a4fe-5aeaa51935f3"
        processor="logger"
        atLeast="#[0]">
        <munit-tools:with-attributes>
          <munit-tools:with-attribute
            whereValue="Some logic for CODE_A"
            attributeName="message" />
        </munit-tools:with-attributes>
      </munit-tools:verify-call>
    </munit:validation>
  </munit:test>
  <!-- -->
  <!-- -->
  <!-- -->
  <flow name="munit-set-error-code-event-flow">
    <logger
      level="TRACE"
      doc:name="doc: Set Event" />
    <!-- [STUDIO:"Generate error with exception = resources/option-d/httpError.dwl"]<munit:set-event
      doc:name="Generate error with exception = resources/option-d/httpError.dwl"
      cloneOriginalEvent="true">
      <munit:error
        id="HTTP:INTERNAL_SERVER_ERROR"
        exception="#[${file::option-d/httpError.dwl}&#93;" />
    </munit:set-event> [STUDIO] -->
  </flow>
  <!-- -->
  <!-- -->
  <!-- -->
  <munit:test
    name="test-test-suite-test-inline-FlowTest"
    timeOut="9000000">
    <munit:enable-flow-sources>
      <munit:enable-flow-source
        value="munit-set-error-code-event-inline-flow" />
    </munit:enable-flow-sources>
    <munit:behavior>
      <munit-tools:mock-when
        doc:name="Mock when Request -&gt; then call"
        processor="http:request">
        <munit-tools:with-attributes>
          <munit-tools:with-attribute
            whereValue="GET"
            attributeName="method" />
          <munit-tools:with-attribute
            whereValue="http://example.com/some-resource"
            attributeName="url" />
        </munit-tools:with-attributes>
        <munit-tools:then-call
          flow="munit-set-error-code-event-inline-flow" />
      </munit-tools:mock-when>
    </munit:behavior>
    <munit:execution>
      <flow-ref
        doc:name="Flow-ref to testFlow"
        name="test2Flow" />
    </munit:execution>
    <munit:validation>
      <munit-tools:verify-call
        doc:name="Some logic for CODE_B"
        doc:id="f5099a83-7189-40d1-abb0-d5f6e228d720"
        processor="logger">
        <munit-tools:with-attributes>
          <munit-tools:with-attribute
            whereValue="Some logic for CODE_B"
            attributeName="message" />
        </munit-tools:with-attributes>
      </munit-tools:verify-call>
    </munit:validation>
  </munit:test>
  <!-- -->
  <!-- -->
  <!-- -->
  <flow name="munit-set-error-code-event-inline-flow">
    <logger
      level="TRACE"
      doc:name="doc: Set Event" />
    <munit:set-event
      doc:name="Generate error with exception = inline java http exception (Doesn't seem to work as expected)"
      cloneOriginalEvent="true">
      <munit:error
        id="HTTP:INTERNAL_SERVER_ERROR"
        exception="#[java!org::mule::extension::http::api::request::validator::ResponseValidatorTypedException::new('Internal server error',  java!org::mule::extension::http::api::error::HttpError::INTERNAL_SERVER_ERROR, java!org::mule::runtime::api::message::Message::of(  java!org::mule::runtime::api::metadata::TypedValue::new( write({errorCode:'CODE_B'},'application/json',{indent:false}), java!org::mule::runtime::api::metadata::DataType::JSON_STRING ) ) )]" />
    </munit:set-event>
  </flow>
</mule>
